### PR TITLE
Update all forge event code, attempt to become loader unaware

### PR DESCRIPTION
**Pull request by Thodor12**
# Changes proposed in this pull request
- Introduce a new event handler in the IMinecoloniesAPI, which is backed by an interface containing basic methods for the events we want to send
- Event handler is implemented in such a way it could be made loader unaware in the future if needed.
- All logic for sending Forge specific events is ripped out in favor of the API event handler

## Testing
- [X] Yes I tested this before submitting it.
- [ ] I also did a multiplayer test.

Review please

### CODE REVIEW

File: src/main/java/com/minecolonies/api/eventbus/DefaultEventBus.java [hunk 1]
```diff
@@ -0,0 +1,54 @@
...
+     */
+    private final Map<TypeToken<?>, List<EventHandler<?>>> eventHandlersPerType = new HashMap<>();
+
+    @Override
+    public <T extends IModEvent> void subscribe(final @NotNull IModEventType<T> eventType, final @NotNull EventHandler<T> handler)
+    {
+        Log.getLogger().debug("Registering event handler for id {}.", eventType.getIdentifier());
+
+        eventHandlersPerType.putIfAbsent(eventType.getIdentifier(), new ArrayList<>());
+        eventHandlersPerType.get(eventType.getIdentifier()).add(handler);
```
Comment (Nightenom) [inline src/main/java/com/minecolonies/api/eventbus/DefaultEventBus.java]: seems like double hashing to me, computeIfAbsent(key).add(value)?

File: src/main/java/com/minecolonies/api/eventbus/DefaultEventBus.java [hunk 2]
```diff
@@ -0,0 +1,54 @@
...
+
+        for (final EventHandler<?> handler : eventHandlers)
+        {
+            try
+            {
+                ((EventHandler<T>) handler).apply(event);
+            }
+            catch (Exception ex)
+            {
+                Log.getLogger().warn("Event id: {};'Error occurred during processing event handler:", event.getEventId(), ex);
```
Comment (Nightenom) [inline src/main/java/com/minecolonies/api/eventbus/DefaultEventBus.java]:
```suggestion
                Log.getLogger().warn("Error occurred during processing event '{}', handler '{}':", event.getEventId(), handler.toString() or handler#class#toString(), ex);
```

File: src/main/java/com/minecolonies/api/eventbus/events/colony/AbstractColonyEvent.java [hunk 3]
```diff
@@ -1,37 +1,36 @@
...
+
+    /**
+     * Gets the colony related to the event.
+     */
+    @NotNull
+    public IColony getColony()
+    {
+        return colony;
+    }
+}
```
Comment (Nightenom) [inline src/main/java/com/minecolonies/api/eventbus/events/colony/AbstractColonyEvent.java]: there are few classes apparently with cr/lf changes

File: src/main/java/com/minecolonies/apiimp/CommonMinecoloniesAPIImpl.java [hunk 4]
```diff
@@ -52,34 +54,36 @@
...
+    private IForgeRegistry<ResearchEffectEntry>                        researchEffectRegistry;
+    private IForgeRegistry<ResearchCostType>                           researchCostRegistry;
+    private IForgeRegistry<RecipeTypeEntry>                            recipeTypeEntryRegistry;
+    private IForgeRegistry<CraftingType>                               craftingTypeRegistry;
+    private IForgeRegistry<QuestRegistries.ObjectiveEntry>             questObjectiveRegistry;
+    private IForgeRegistry<QuestRegistries.RewardEntry>                questRewardRegistry;
+    private IForgeRegistry<QuestRegistries.TriggerEntry>               questTriggerRegistry;
+    private IForgeRegistry<QuestRegistries.DialogueAnswerEntry>        questDialogueAnswerRegistry;
+    private IForgeRegistry<HappinessRegistry.HappinessFactorTypeEntry> happinessFactorTypeRegistry;
+    private IForgeRegistry<HappinessRegistry.HappinessFunctionEntry>   happinessFunctionRegistry;
```
Comment (Nightenom) [inline src/main/java/com/minecolonies/apiimp/CommonMinecoloniesAPIImpl.java]: I mean can we generally avoid unnecessary whitespace changes? _or ideally finally make proper autoformat_ :D

**Comment by Nightenom** [comment]
Also (without looking at code if it's even possible) cannot we reuse existing forge eventbus for implementing the bus logic?

File: src/main/java/com/minecolonies/apiimp/CommonMinecoloniesAPIImpl.java (hunk 4 above)
Comment (Thodor12) [reply src/main/java/com/minecolonies/apiimp/CommonMinecoloniesAPIImpl.java]: I mean we technically have a format guideline, it just wasn't applied to this file :D

**Comment by Thodor12** [comment]
> Also (without looking at code if it's even possible) cannot we reuse existing forge eventbus for implementing the bus logic?

The point was from the beginning to not use Forge code where possible, so we quickly glanced over using a Guava event bus.
But I found that to be less than ideal since it's a third party library Mojang happened to insert, if that were to be dropped we'd have to fallback to shading it, which doesn't work with Guava.

For that reason I thought it was a better idea to make a small lightweight event bus tailored to our needs.

File: src/main/java/com/minecolonies/api/eventbus/events/colony/AbstractColonyEvent.java (hunk 3 above)
Comment (Thodor12) [reply src/main/java/com/minecolonies/api/eventbus/events/colony/AbstractColonyEvent.java]: These are just weird diffs by Github, it's because the file moved and slightly altered it doesn't know what to do with them.

**Comment by someaddons** [comment]
Forge eventbus is an extion of guavas afaik

**Comment by uecasm** [comment]
What's the point of avoiding (Neo)Forge APIs?  Unless you're anticipating a Fabric port in the future or something, but that can be handled with an abstraction layer rather than a different implementation.

File: src/main/java/com/minecolonies/api/eventbus/events/colony/AbstractColonyEvent.java (hunk 3 above)
Comment (Raycoms) [reply src/main/java/com/minecolonies/api/eventbus/events/colony/AbstractColonyEvent.java]: Yeah, but we shouldn't auto format all files we touch, because that makes it impossible to read the dif

Comment (Raycoms) [reply src/main/java/com/minecolonies/api/eventbus/events/colony/AbstractColonyEvent.java]: Also applies to the other completely formatted files. Aside from porting madness problems =D

File: src/main/java/com/minecolonies/core/entity/citizen/EntityCitizen.java [hunk 5]
```diff
@@ -1627,18 +1628,18 @@ public void die(@NotNull final DamageSource damageSource)
...
 
-            try
-            {
-                MinecraftForge.EVENT_BUS.post(new CitizenRemovedEvent(citizenData, damageSource));
-            }
-            catch (final Exception e)
-            {
-                Log.getLogger().error("Error during CitizenRemovedEvent", e);
-            }
+            IMinecoloniesAPI.getInstance().getEventBus().post(MinecoloniesEventTypes.CITIZEN_DIED, new com.minecolonies.api.eventbus.events.colony.citizens.CitizenDiedEvent(citizenData, damageSource));
```
Comment (Raycoms) [inline src/main/java/com/minecolonies/core/entity/citizen/EntityCitizen.java]: too long import

File: src/main/java/com/minecolonies/core/colony/ColonyManager.java [hunk 6]
```diff
@@ -202,7 +204,7 @@ private void deleteColony(@Nullable final IColony iColony, final boolean canDest
                 return;
             }
 
-            ColonyEvents.deleteColony(colony);
+            IMinecoloniesAPI.getInstance().getEventBus().post(MinecoloniesEventTypes.COLONY_DELETED, new ColonyDeletedEvent(colony));
```
Comment (Raycoms) [inline src/main/java/com/minecolonies/core/colony/ColonyManager.java]: Seems a bit long. Couldn't we
a) Do a singleton: DefaultEventBus.post
b) just post the new ColonyDeletedEvent, without needing the type constant?

**Review by Raycoms — CHANGES_REQUESTED** [review]
Two smaller comments

**Review by Raycoms — APPROVED** [review]

File: src/main/java/com/minecolonies/api/eventbus/DefaultEventBus.java [hunk 7]
```diff
@@ -0,0 +1,52 @@
...
+            return;
+        }
+
+        Log.getLogger().debug("Sending event '{}' for type '{}'. Sending to {} handlers.", event.getEventId(), event.getClass().getSimpleName(), eventHandlers.size());
+
+        for (final EventHandler<? extends IModEvent> handler : eventHandlers)
+        {
+            try
+            {
+                ((EventHandler<IModEvent>) handler).apply(event);
```
Comment (Nightenom) [inline src/main/java/com/minecolonies/api/eventbus/DefaultEventBus.java:44]: Mby change post also to `? extends Event` to avoid typing errors

**Review by Nightenom — COMMENTED** [review]
Generally speaking I would avoid renaming event classes to ModEvent, just imho though

File: src/main/java/com/minecolonies/api/eventbus/DefaultEventBus.java (hunk 7 above)
Comment (Thodor12) [reply src/main/java/com/minecolonies/api/eventbus/DefaultEventBus.java:44]: That doesn't matter, for one `? extends Event` is not valid for arguments, that's not valid syntax, only for type arguments.
Also the problem lies in the fact that Java cannot match the type argument for the key and value of the map to be identical. So even if I were to put a generic type on it, it won't be able to match it.
Lastly I have to do this cast because a method which takes an argument of `? extends Event` does not accept `Event` itself, because java dumb.

**Review by someaddons — APPROVED** [review]
